lib/web_ui/lib/src/engine/canvaskit/image_web_codecs.dart

```diff
@@ -0,0 +1,384 @@
+// found in the LICENSE file.
+
+/// Uses the `ImageDecoder` class supplied by the browser.
+///
+/// See also:
+///
+///  * `image_wasm_codecs.dart`, which uses codecs supplied by the CanvasKit WASM bundle.
+@JS()
+library image_web_codecs;
+
+import 'dart:async';
+import 'dart:js_util' as js_util;
+import 'dart:math' as math;
+import 'dart:typed_data';
+
+import 'package:js/js.dart';
+import 'package:ui/ui.dart' as ui;
+
+import 'canvaskit_api.dart';
+import 'image.dart';
+import 'surface.dart';
+import 'surface_factory.dart';
+
+@JS('window.ImageDecoder')
+external Object? get _imageDecoderConstructor;
+
+/// Whether the current browser supports `ImageDecoder`.
+bool browserSupportsImageDecoder =
+    _imageDecoderConstructor != null;
+
+/// Sets the value of [browserSupportsImageDecoder] to its default value.
+void debugResetBrowserSupportsImageDecoder() {
+  browserSupportsImageDecoder =
+      _imageDecoderConstructor != null;
+}
+
+/// Image decoder backed by the browser's `ImageDecoder`.
+class CkBrowserImageDecoder implements ui.Codec {
+  static Future<CkBrowserImageDecoder> create({
+    required Uint8List data,
+    required String debugSource,
+    int? targetWidth,
+    int? targetHeight,
+  }) async {
+    // ImageDecoder does not detect image type automatically. It requires us to
+    // tell it what the image type is.
+    final String? contentType = detectContentType(data);
+
+    if (contentType == null) {
+      throw ImageCodecException(
+        'Failed to detect image file format of the image data using the file header, which was:\n' +
+        data.sublist(0, math.min(10, data.length)).map((int byte) => byte.toRadixString(16).padLeft(2, '0')).join(' '),
+      );
+    }
+
+    if (!await _isImageTypeSupported(contentType)) {
+      throw ImageCodecException(
+        'Image file format $contentType is not supported by this browser\'s ImageDecoder.',
+      );
+    }
+
+    final _ImageDecoder webDecoder = _ImageDecoder(_ImageDecoderOptions(
+      type: contentType,
+      data: data,
+
+      // Flutter always uses premultiplied alpha.
+      premultiplyAlpha: 'premultiply',
+      desiredWidth: targetWidth,
+      desiredHeight: targetHeight,
+
+      // "default" gives the browser the liberty to convert to display-appropriate
+      // color space, typically SRGB, which is what we want.
+      colorSpaceConversion: 'default',
+
+      // Flutter doesn't give the developer to customize this, so if this is an
+      // animated image we should prefer the animated track.
+      preferAnimation: true,
+    ));
+
+    await js_util.promiseToFuture<void>(webDecoder.tracks.ready);
+
+    // Flutter doesn't have an API for progressive loading of images, so we
+    // wait until the image is fully decoded.
+    // package:js bindings don't work with getters that return a Promise, which
+    // is why js_util is used instead.
+    await js_util.promiseToFuture<void>(js_util.getProperty(webDecoder, 'completed'));
```

Comment:
completed is for loading, not decoding. No decoding starts until decode() is called. This will ensure the frame count is accurate though.